### ISSUE TITLE
participantDidDisconnect is being called too late

### ISSUE THREAD

**Issue by vysotsky** [twilio/video-quickstart-swift#132]
Hi guys! Is there an opportunity to check if user just got disconnected?
Because participantDidDisconnect method is called after 1-2 minutes after interlocutor has been disconnected due to internet issues etc. This means that all this time I see a freezed TVIVideoView instead of a message saying that something went wrong etc.
Thanks!

**Comment by vysotsky** [comment]
There is another problem - when interlocutor gets trouble connecting but restores it after a while then TVIVideoView stops showing video stream and delegate is not being called. I have no idea what's going on.

**Comment by ceaglest** [comment]
Hi @vysotsky,

> Hi guys! Is there an opportunity to check if user just got disconnected?
Because participantDidDisconnect method is called after 1-2 minutes after interlocutor has been disconnected due to internet issues etc. This means that all this time I see a freezed TVIVideoView instead of a message saying that something went wrong etc.

This is a known issue, please see: https://github.com/twilio/video-quickstart-swift/issues/99

> There is another problem - when interlocutor gets trouble connecting but restores it after a while then TVIVideoView stops showing video stream and delegates are not being called. I have no idea what's going on.

Once your signaling connection has been restored, the Client should detect the media connectivity failure and restore the media connection. Are you saying that this isn't happening, or that its taking longer than expected? If so, please provide:

1. Twilio Video version
2. Device / OS version
3. Debug logs showing the problem

In addition to the signaling timeout mentioned in #99, we are also planning on making improvements to detecting media connectivity errors (fail faster, restore sooner).

Regards,
Chris Eagleston

**Comment by vysotsky** [comment]
Hi @ceaglest, thanks for the quick reply!

I'm saying that the client no longer shows stream after connection on the other device was interrupted. I'm using two devices for testing. First one is iPod touch (6th generation) with iOS 10.3 on it, second one is iPod touch (5th generation) with iOS 9.3.5. 
TwilioVideo 1.0.1

When I disabled Internet on the first iPod 6th generation, there was a lot of warnings like the following:
```
WARN:TwilioVideo:[Signaling]:RESIP::TRANSPORT: Can't find matching transport [ V4 127.0.0.1:0 TLS target domain=unspecified mFlowKey=0 ]
WARN:TwilioVideo:[Signaling]:RESIP::TRANSPORT: Can't find matching transport [ V4 127.0.0.1:0 TLS target domain=unspecified mFlowKey=0 ]
WARN:TwilioVideo:[Signaling]:RESIP::TRANSPORT: Can't find matching transport [ V4 10.0.4.148:0 TCP target domain=unspecified mFlowKey=0 ]
WARN:TwilioVideo:[Core]:No info for track
WARN:TwilioVideo:[Core]:No info for track
WARN:TwilioVideo:[Core]:No info for track
WARN:TwilioVideo:[Core]:No info for track
WARN:TwilioVideo:[Signaling]:RESIP::STATS: RESIP:TRANSACTION
TU summary: 0 TRANSPORT 0 TRANSACTION 0 CLIENTTX 2 SERVERTX 0 TIMERS 7
Transaction summary: reqi 8 reqo 20 rspi 16 rspo 8
Details: INVi 0/S0/F0 INVo 4/S3/F1 ACKi 0 ACKo 4 BYEi 0/S0/F0 BYEo 0/S0/F0 CANi 0/S0/F0 CANo 0/S0/F0 MSGi 0/S0/F0 MSGo 0/S0/F0 OPTi 0/S0/F0 OPTo 0/S0/F0 REGi 0/S0/F0 REGo 0/S0/F0 PUBi 0/S0/F0 PUBo 0/S0/F0 SUBi 0/S0/F0 SUBo 0/S0/F0 NOTi 0/S0/F0 NOTo 0/S0/F0 REFi 0/S0/F0 REFo 0/S0/F0 INFi 8/S8/F0 INFo 12/S12/F0 PRAi 0/S0/F0 PRAo 0/S0/F0 SERi 0/S0/F0 SERo 0/S0/F0 UPDi 0/S0/F0 UPDo 0/S0/F0
Retransmissions: INVx 0 finx 0 nonx 0 BYEx 0 CANx 0 MSGx 0 OPTx 0 REGx 0 PUBx 0 SUBx 0 NOTx 0 REFx 0 INFx 0 PRAx 0 SERx 0 UPDx 0
```
Then i turned on the Internet connection on the first device and as a result second device restored stream for a while and after a few seconds it stoped working again. After this I was not able to restore the stream anymore.

BTW, I cannot find any fresh information what devices are compatible with the new Video Client.  [This article](https://support.twilio.com/hc/en-us/articles/223180828-What-devices-are-compatible-with-the-Twilio-Client-iOS-SDK) says that iPod of 5th generation is one of the supported devices, it has a front camera, but the Video Client cannot allocate a Camera capturer for the device.

`ERROR:TwilioVideo:[Platform]:Failed to create camera capturer, source not available.`

**Comment by ceaglest** [comment]
Hi @vysotsky,

> BTW, I cannot find any fresh information what devices are compatible with the new Video Client. This article says that iPod of 5th generation is one of the supported devices, it has a front camera, but the Video Client cannot allocate a Camera capturer for the device.

TwilioVideo does not officially support the [iPod Touch (5th generation)](https://www.twilio.com/docs/api/video/download-video-sdks#supported-devices_1). The link you mentioned is for Twilio Client which does support A5 based devices. We don't do anything to prevent the SDK for running on the A5, but you shouldn't expect a very good experience on an iPod Touch 5G. As for the camera, some variants of the iPod Touch 5G have a front & rear camera while others have just a rear camera. `TVICameraCapturer` chooses the front camera when you use the default initializer. Can you use the class methods `isSourceAvailable` and `availableSources` to determine what is available on your device?

> Then i turned on the Internet connection on the first device and as a result second device restored stream for a while and after a few seconds it stoped working again. After this I was not able to restore the stream anymore.

Can you provide debug logs for both devices before, during, and after network connectivity is lost? This will help me to determine what is going on. As I mentioned our SDK should attempt to automatically restore the signaling and media connections once network connectivity returns.

Best,
Chris

**Comment by chublix** [comment]
Hi @ceaglest, @vysotsky can't reply you now. That's logs for both devices
[log_06-02-2017_05-33-28.txt](https://github.com/twilio/video-quickstart-swift/files/1055269/log_06-02-2017_05-33-28.txt)
[log_06-02-2017_05-33-28.txt](https://github.com/twilio/video-quickstart-swift/files/1055270/log_06-02-2017_05-33-28.txt)

**Comment by ceaglest** [comment]
Hi @chublix @vysotsky,

Sorry for the delay in responding. It looks like these logs are both from the same device. However, I grabbed the Room SID to see what was going on. It looks like you've encountered an issue with network handoff in a Group Room. At the moment your media connection to our server is not restored if you lose your signaling connection, and then restore it. We are working to fix this during our Group Rooms beta period.

So to summarize the issues you've encountered:

Participant disconnect called late: 
See: https://github.com/twilio/video-quickstart-swift/issues/99

iPod Touch 5G: 
Not officially supported see my message above to fix TVICameraCapturer initialization.

Network Handoff for Media in Group Rooms:
Does not work right now, we are planning on fixing this during the Group Rooms beta.

Regards,
Chris Eagleston

**Comment by ptankTwilio** [comment]
@vysotsky @chublix  I have created a new issue #[154](https://github.com/twilio/video-quickstart-swift/issues/154) for network handoff for media in group rooms. I would recommend following it as we are planning to fix the hand over issue soon.

Closing the ticket as the problems reported on this tickets are covered by #[99](https://github.com/twilio/video-quickstart-swift/issues/99) and #[154](https://github.com/twilio/video-quickstart-swift/issues/154).
